planner/core/collect_column_stats_usage_test.go

```diff
@@ -12,210 +12,302 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-package core_test
+package core
 
 import (
 	"context"
 	"fmt"
-	"testing"
+	"sort"
 
+	. "github.com/pingcap/check"
+	"github.com/pingcap/tidb/infoschema"
 	"github.com/pingcap/tidb/parser/model"
-	plannercore "github.com/pingcap/tidb/planner/core"
-	"github.com/pingcap/tidb/testkit"
 	"github.com/pingcap/tidb/util/hint"
-	"github.com/pingcap/tidb/util/logutil"
-	"github.com/stretchr/testify/require"
+	"github.com/pingcap/tidb/util/testleak"
 )
 
-func TestCollectPredicateColumns(t *testing.T) {
-	store, dom, clean := testkit.CreateMockStoreAndDomain(t)
-	defer clean()
-	tk := testkit.NewTestKit(t, store)
-	tk.MustExec("use test")
-	tk.MustExec("drop table if exists t1, t2")
-	tk.MustExec("set @@session.tidb_partition_prune_mode = 'static'")
-	tk.MustExec("create table t1(a int, b int, c int)")
-	tk.MustExec("create table t2(a int, b int, c int)")
-	tk.MustExec("create table t3(a int, b int, c int) partition by range(a) (partition p0 values less than (10), partition p1 values less than (20), partition p2 values less than maxvalue)")
+func getColumnName(c *C, is infoschema.InfoSchema, tblColID model.TableColumnID, comment CommentInterface) (string, bool) {
+	var tblInfo *model.TableInfo
+	var prefix string
+	if tbl, ok := is.TableByID(tblColID.TableID); ok {
+		tblInfo = tbl.Meta()
+		prefix = tblInfo.Name.L + "."
+	} else {
+		db, exists := is.SchemaByName(model.NewCIStr("test"))
+		c.Assert(exists, IsTrue, comment)
+		for _, tbl := range db.Tables {
+			pi := tbl.GetPartitionInfo()
+			if pi == nil {
+				continue
+			}
+			for _, def := range pi.Definitions {
+				if def.ID == tblColID.TableID {
+					tblInfo = tbl
+					prefix = tbl.Name.L + "." + def.Name.L + "."
+					break
+				}
+			}
+			if tblInfo != nil {
+				break
+			}
+		}
+		if tblInfo == nil {
+			return "", false
+		}
+	}
+	for _, col := range tblInfo.Columns {
+		if tblColID.ColumnID == col.ID {
+			return prefix + col.Name.L, true
+		}
+	}
+	return "", false
+}
 
+func checkColumnStatsUsage(c *C, is infoschema.InfoSchema, lp LogicalPlan, onlyHistNeeded bool, expected []string, comment CommentInterface) {
+	var tblColIDs []model.TableColumnID
+	if onlyHistNeeded {
+		_, tblColIDs = CollectColumnStatsUsage(lp)
+	} else {
+		tblColIDs, _ = CollectColumnStatsUsage(lp)
+	}
+	cols := make([]string, 0, len(tblColIDs))
+	for _, tblColID := range tblColIDs {
+		col, ok := getColumnName(c, is, tblColID, comment)
+		c.Assert(ok, IsTrue, comment)
+		cols = append(cols, col)
+	}
+	sort.Strings(cols)
+	c.Assert(cols, DeepEquals, expected, comment)
+}
+
+func (s *testPlanSuite) TestCollectPredicateColumns(c *C) {
```

Comment:
Please add more test cases for `prepare` statement, same to `TestCollectHistNeededColumns`.

Reply:
Besides, add some test cases for `subquery`.

Reply:
More tests about `subquery` have been added for both `TestCollectPredicateColumns` and `TestCollectHistNeededColumns`. In `collect_column_stats_usage_test.go`(in `planner/core` package) we cannot run `tk.MustExec("prepare ...")` because it is unit test and uses mock(For similar reason I don't test non-correlated scalar subquery here, which executes the subquery). Usually we have to run integration tests with `tk.MustExec` in `planner/core_test` package to avoid circle import. I will add integration tests to test `prepare` and non-correlated scalar subquery in #30816.